File: Java/api/treehopper.libraries/src/main/java/io/treehopper/libraries/displays/Led.java
package io.treehopper.libraries.displays;

/**
 * LED
 */

public class Led {

    private ILedDriver driver;
    private boolean brightnessControl;
    int channel;
    private boolean state = false;
    private double brightness = 0.0;

    public Led(ILedDriver driver, int channel) {
        this(driver, channel, false);
    }

    public Led(ILedDriver driver, int channel, boolean hasBrightnessControl) {
        this.driver = driver;
        this.brightnessControl = hasBrightnessControl;
        this.channel = channel;
    }

    public int getChannel() {
        return channel;
    }

    public ILedDriver getDriver() {
        return driver;
    }

    public double getBrightness() {
        return brightness;
    }
    

    public void setBrightness(double value) {
        if (brightness == value) return;
        if (!hasBrightnessControl()) return;
        brightness = value;
        driver.ledBrightnessChanged(this);
    }


    public boolean hasBrightnessControl() {
        return brightnessControl;
    }


    public boolean getState() {
        return state;
    }

    public void setState(boolean value) {
        if (state == value) return;
        state = value;
        driver.ledStateChanged(this);
    }
}

File: C++/API/Treehopper.Libraries/src/Displays/Led.cpp
#include "Utility.h"
#include "Displays/Led.h"
#include "Displays/LedDriver.h"
namespace Treehopper {
	namespace Libraries {
		namespace Displays {
			using namespace Treehopper;
			Led::Led(LedDriver& driver, int channel, bool hasBrightnessControl) : driver(driver), channel(channel), hasBrightnessControl(hasBrightnessControl)
			{
			}
			void Led::brightness(double value)
			{
				if (Utility::closeTo(value, _brightness)) return;
				if (!hasBrightnessControl) return;
				_brightness = value;
				driver.ledBrightnessChanged(*this);

				if (driver.autoFlush)
					driver.flush();
			}
			double Led::brightness()
			{
				return _brightness;
			}
			bool Led::state()
			{
				return _state;
			}
			void Led::state(bool value)
			{
				if (_state == value) return;
				_state = value;
				driver.ledStateChanged(*this);

				if (driver.autoFlush)
					driver.flush();
			}
		}
	}
}
